core/src/impl/Kokkos_ViewHookDeepCopy.hpp

```diff
@@ -0,0 +1,134 @@
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_VIEWHOOK_DEEPCOPY_HPP
+#define KOKKOS_VIEWHOOK_DEEPCOPY_HPP
+
+#include <Kokkos_Core_fwd.hpp>
+#include <functional>
+#include <memory>
+#include <type_traits>
+
+namespace Kokkos {
+namespace Experimental {
+namespace Impl {
+/**
+  * Here we define copy_buffer utility functions
+  * to copy data to and from Views. We need
+  * specialized variants for ScratchMemorySpace
+  * and AnonymousSpace as they don't participate
+  * in the ViewHooks system, and we need copies
+  * to and from Views
+  */
+template <class ViewType>
+typename std::enable_if<(std::is_same<Kokkos::AnonymousSpace,
+                                      typename ViewType::memory_space>::value ||
+                         std::is_same<Kokkos::ScratchMemorySpace<
+                                          typename ViewType::execution_space>,
+                                      typename ViewType::memory_space>::value),
+                        void>::type
+copy_buffer(unsigned char *, ViewType &) {}
+
+template <class ViewType>
+typename std::enable_if<
+    !(std::is_same<Kokkos::AnonymousSpace,
+                   typename ViewType::memory_space>::value ||
+      std::is_same<
+          Kokkos::ScratchMemorySpace<typename ViewType::execution_space>,
+          typename ViewType::memory_space>::value),
+    void>::type
+copy_buffer(unsigned char *buff, ViewType &v) {
+  Kokkos::Impl::DeepCopy<Kokkos::HostSpace, typename ViewType::memory_space,
+                         typename ViewType::execution_space>(
+      (void *)buff, v.data(), v.span() * sizeof(typename ViewType::value_type));
+}
+
+template <class ViewType>
+typename std::enable_if<(std::is_same<Kokkos::AnonymousSpace,
+                                      typename ViewType::memory_space>::value ||
+                         std::is_same<Kokkos::ScratchMemorySpace<
+                                          typename ViewType::execution_space>,
+                                      typename ViewType::memory_space>::value ||
+                         std::is_const<typename ViewType::value_type>::value),
+                        void>::type
+copy_buffer(ViewType &, unsigned char *) {}
+
+template <class ViewType>
+typename std::enable_if<
+    !(std::is_same<Kokkos::AnonymousSpace,
+                   typename ViewType::memory_space>::value ||
+      std::is_same<
+          Kokkos::ScratchMemorySpace<typename ViewType::execution_space>,
+          typename ViewType::memory_space>::value ||
+      std::is_const<typename ViewType::value_type>::value),
+    void>::type
+copy_buffer(ViewType &v, unsigned char *buff) {
+  Kokkos::Impl::DeepCopy<typename ViewType::memory_space, Kokkos::HostSpace,
+                         typename ViewType::execution_space>(
+      v.data(), (void *)buff, v.span() * sizeof(typename ViewType::value_type));
+}
+
+template <class ViewType>
+class ViewHookDeepCopy<ViewType, void> {
```

Comment:
what is the purpose of this class?

Reply:
why can't a deep_copy to a buffer not be done with a proper deep_copy to an unmanaged view?

Reply:
the separation between DeepCopy<> and deep_copy was done mainly because of MAJOR difficulties getting anything that might include ScratchMemorySpace to work with deep_copy.  If there is a good example of how to make that work in this context, I would be glad to integrate it.

Reply:
Can you explain "MAJOR" difficulties? I'm skeptical also

Reply:
merged/resolved in 9dcc4ccbd64596b95753d8996a19c1c3e8929739